Fix Sys.file_exists and Sys.is_directory for symlinks on Windows

dra27 opened the pull request:
While poking around in `runtime/sys.c` for #9787, I spotted serious problems with `caml_sys_file_exists` and `caml_sys_is_directory` - they use `stat`, which is a symlink catastrophe on Windows.

This PR makes the use of `stat_os` in the codebase an error if `CAML_INTERNALS` is defined (I'm not sure why the `_os` functions aren't guarded with `CAML_INTERNALS`, @nojb?). The function is so unreliable on Windows that it should simply never be used - the idea with the error is to sound an alarm bell that a Windows-specific piece of code will be needed.

My solution is to define a function which reads the attributes of the file entry. This - like `stat` on Unix - is done without actually opening the file. I decided to check how our friends over in Redmond do it in `FileSystem.FileExists` (it's much easier than in 2015, since .NET is open source now!) which alerted me to the matter of zombie files. These are files which have been deleted but which are still open in programs (this is real - both ocamltest and dune have had visible issues with this). Clearly `Sys.file_exists` should return `true` if you can still "see" a file with the name given - the fact you can't actually open it is a security/permissions issue. Fortunately, when this happens `GetLastError` returns one of three known errors and so in this specific case we then use the much slower `FindFirstFile` which - in what can only be described as Windows weirdness - does return the file attributes, returning a reliable answer for `Sys.file_exists` (I don't actually know if you can concoct zombie directories - you can't in directly in OCaml).

This made me realise that my implementation of `stat` in win32unix which is based on `GetFileInformationByHandle` would therefore fail for these zombie files (Microsoft's broken implementations of stat are all based on `FindFirstFile`). So this PR also fixes `Unix.stat` to return meaningful enough information for files marked for deletion.

Then I realised that the Unix implementation of `caml_sys_file_exists` is based on `stat` but `GetFileAttributes` is an `lstat` like function. This means that on Windows `Sys.file_exists` would return `true` for symlinks which point to files/directories which don't exist. It took a while to convince myself, but I believe that we definitely want `stat`-like not `lstat`-like behaviour for symbolic links. The solution for this is to to follow the symlink by opening the file and using `GetFileInformationByHandle`. Technically speaking, there is more work to do for symbolic links to zombie files - this, however, is quite tricky, very niche, and related to other work, so I leave that for another day.

I have added tests for all of these cases.

dra27 commented on otherlibs/win32unix/stat.c:
```diff
@@ -199,13 +200,52 @@ static int safe_do_stat(int do_lstat, int use_64, wchar_t* path, HANDLE fstat, _
+        info.ftLastAccessTime.dwHighDateTime = alt_info.ftLastAccessTime.dwHighDateTime;
+        info.ftLastWriteTime.dwLowDateTime = alt_info.ftLastWriteTime.dwLowDateTime;
+        info.ftLastWriteTime.dwHighDateTime = alt_info.ftLastWriteTime.dwHighDateTime;
+        /* XXX Could possibly do better with GetVolumeInformation */
+        info.dwVolumeSerialNumber = 0;
+        info.nFileSizeHigh = alt_info.nFileSizeHigh;
+        info.nFileSizeLow = alt_info.nFileSizeLow;
+        /* XXX Could possibly do better with FindFirstFileName */
+        info.nNumberOfLinks = 1;
```
Microsoft's various implementation of `stat` return inconsistent values for `st_dev` (IIRC some return the index of the drive letter, some don't) and `st_nlinks` is always 1. It seems far too involved to loop over `FindFirstFileName` to get the number of hardlinks and translate the path to get the volume to get its serial number for this corner case!

nojb commented:
> I'm not sure why the `_os` functions aren't guarded with `CAML_INTERNALS`, @nojb?

Good question. I can't think of any reason right now, maybe it was just an oversight.

xavierleroy commented:
I'll need more time to digest all this information :-)  Right now here is my 30000-foot view of the situation.

* I agree that `Unix.stat` and `Unix.fstat` should be emulated the best we can on Win32, and that justifies further work on your already admirable code in otherlibs/win32unix.
* I just learned about "zombie files" and I'm completely unsure whether they should be treated as existing or not existing (nice metaphysical question...).  Looks like a case can be made for either interpretation.
* My hope would be to keep `Sys.file_exists` and `Sys.is_directory` as simple as possible under Win32.  With all my admiration for your Win32 mastery, the proposed code feels a bit overengineered.  In ancient times, `Sys.file_exists` was introduced to speed up searching for a file in a search path, instead of trying to open the file then closing it.  Likewise, `Sys.is_directory` was meant as a faster and more intuitive alternative to just trying to read its contents with `Sys.readdir`.  If the simplest Win32 implementation is just that (open and close the file, or read directory contents), maybe that's good enough.  To be discussed, obviously!
* I feel guilty for having used `stat_os` in `runtime/dynlink.c`, function `parse_ld_conf`, and could atone by rewriting the function.

dra27 commented:
There's no rush for this, obviously!

For zombie files, there's a regression if we _stop_ treating them as existing (it's what presently happens). I agree you can argue it both ways, but I'd struggle to hold-my-own convincing someone that the file they can see in Explorer and in a shell prompt `ls`/`dir` does not in fact "exist". The data can still be extracted from it by any process which has the file open!

The primary motivation for the changes to `Sys.file_exists` and `Sys.is_directory` is that they don't work correctly with symbolic links, which I would argue needs to be the case. I have at least with those primitives kept the idea of them the same - the Windows version is - as on Unix - still trying to get the `st_mode` field of a `stat` call. What it's possibly showing is that the reimplementation of `stat` sort of belongs in the runtime (well, on top of the broken CRT underneath our runtime, if you see what I mean!). But I don't want to do that...

I certainly don't think you should feel guilt for using `stat_os`! Indeed, I've removed its uses everywhere for belt and braces - I cannot imagine why you'd actually choose to symlink ld.conf! Of course, you _might_ do that for a dynlink'd library, which would be a puzzling error.

dra27 commented:
Just for comparison, the equivalent .NET Framework function is `System.IO.File.Exists` which is in [runtime/src/libraries/System.IO.FileSystem/src/System/IO/File.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/System.IO.FileSystem/src/System/IO/File.cs#L117-L144); although the more interesting bit of it is in [runtime/src/libraries/Common/src/System/IO/FileSystem.Attributes.Windows.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/Common/src/System/IO/FileSystem.Attributes.Windows.cs#L37-L46) and in particular the internal [`FillAttributeInfo` function](https://github.com/dotnet/runtime/blob/master/src/libraries/Common/src/System/IO/FileSystem.Attributes.Windows.cs#L55-L130). As a side-note, I think that their use of `GetFileAttributesEx` is an Overly-Object-Oriented oversight in that they have a class wrapping that and have forgotten to use the cheaper `GetFileAttributes` call (at least I think it's that, rather than internally `GetFileAttributes` simply calls the `GetFileAttributesEx`, but Microsoft isn't yet open source enough to check that!)

dra27 committed:
Handle symlinks and zombie files on Windows

Sys.is_directory and Sys.file_exists were using stat on Windows which
means that they are affected by bugs in the underlying CRT call. stat is
now used in Unix only, and stat_os updated to be an error if used in the
codebase.

In changing the handling, extra care is required to handle files marked
for deletion on Windows. Files on Windows can be unlinked while open in
other programs, but the file entry remains in the directory until all
the handles to it have been closed. However, these files cannot be opened
or queried directly (hence the "zombie" moniker) - this requires some
careful handling.
